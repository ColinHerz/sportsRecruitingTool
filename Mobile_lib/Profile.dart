import 'package:flutter/material.dart';
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'models/UserInfo.dart';
import 'Session.dart';
import 'ShowDialog.dart';

class Profile extends StatefulWidget {
  @override
  _ProfileState createState() => _ProfileState();
}

class _ProfileState extends State<Profile> {

  // Keyboard focus nodes for each field
  List<FocusNode> _nodes = [new FocusNode(), new FocusNode(), new FocusNode()];

  // ReadOnly booleans for each field
  List<bool> _ReadOnlybools = [true, true, true];

  // Text controllers for the profile information fields
  final _nameController = TextEditingController(text: "$SessionFirstName $SessionLastName");
  final _emailController = TextEditingController(text: SessionEmail);
  final _ageController = TextEditingController(text: SessionAge.toString());
  final _heightController = TextEditingController(text: SessionHeight.toString());
  final _weightController = TextEditingController(text: SessionWeight.toString());

  bool isVerified = false;

  // Dialog box
  void _showDialog(String message, int enable){
    showDialog(
        context: context,
        builder: (BuildContext context){
          return AlertDialog(
              title: new Text(message),
              actions: <Widget>[
                new FlatButton(
                    child: new Text("Yes"),
                    onPressed: (){
                      _ReadOnlybools[enable] = false;
                      Navigator.of(context).pop();
                    }
                ),
                new FlatButton(
                  child: new Text("Cancel"),
                  onPressed: (){
                    _ReadOnlybools[enable] = true;
                    Navigator.of(context).pop();
                    },
                )
              ]
          );
        }
    );
  }

  Future<UserInfo> postUser(dynamic info) async {
    final http.Response _response = await http.post(
      'http://localhost3000.us-east-2.elasticbeanstalk.com/api/users/detail/update',
      headers: headers,
      body: info,
    );

    // If successful login, navigate to homepage.
    if (_response.statusCode == 200) {
      return UserInfo.fromJson(json.decode(_response.body));
    }
    else {
      print(_response.statusCode);
      print(_response.body);
    }
  }
  Future<UserInfo> getUser() async {
    final http.Response _response = await http.get(
      'http://localhost3000.us-east-2.elasticbeanstalk.com/api/users/get',
      headers: headers,
    );

    // If successful login, navigate to homepage.
    if (_response.statusCode == 200) {
      var parsedJson = json.decode(_response.body);
      setState(() {
        isVerified = parsedJson['isVerified'];
      });
      //return UserInfo.fromJson(json.decode(_response.body));
    }
    else {
      print(_response.statusCode);
      print(_response.body);
    }
  }
  Future<UserInfo> resendVerification(dynamic info) async {
    final http.Response _response = await http.post(
      'http://localhost3000.us-east-2.elasticbeanstalk.com/api/emails/resendVerificationEmail',
      headers: headers,
      body: info,
    );

    // If successful login, navigate to homepage.
    if (_response.statusCode == 200) {
      DialogPopUp(context, "Email has been re-sent!");
      //return UserInfo.fromJson(json.decode(_response.body));
    }
    else {
      print(_response.statusCode);
      print(_response.body);
    }
  }

  InkWell isNotVerified(){
    return InkWell(
      child: Text(
        "Need to resend verification email?",
        style: TextStyle(
            fontSize: 16,
            color: Colors.deepOrangeAccent
        ),
      ),
      onTap: (){
        resendVerification(jsonEncode(<String, String>{
          'receiverEmail' : _emailController.text
        }));
        },
    );
  }

  void initState()
  {
    super.initState();
    getUser();
  }

  Widget build(BuildContext context){
    return Scaffold(
      appBar: AppBar(
          title: Text('Profile')
      ),
      body: ListView(
          children: <Widget>[
        Column(
          mainAxisAlignment: MainAxisAlignment.start,
            children: <Widget>[
              Padding(
                padding: EdgeInsets.fromLTRB(0, 10, 0, 10),
                child: TextField(
                    readOnly: true,
                    controller: _nameController,
                    decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    labelText: 'Name',
                  )
                )
              ),
              Padding(
                  padding: EdgeInsets.fromLTRB(0, 0, 0, 10),
                  child: TextField(
                      readOnly: true,
                      controller: _emailController,
                      decoration: InputDecoration(
                        border: OutlineInputBorder(),
                        labelText: 'Email',
                      )
                  )
              ),
              Padding(
                padding: EdgeInsets.fromLTRB(0, 0, 0, 10),
                child: Text("Email verified: $isVerified",
                    style: TextStyle(
                        fontSize: 20,
                        fontWeight: FontWeight.bold,
                        color: Colors.deepOrange
                    )
                ),
              ),
              isNotVerified(),
              Padding(
                  padding: EdgeInsets.fromLTRB(0, 10, 0, 10),
                  child: TextField(
                      onTap: () {
                        setState(() {
                          _showDialog("Edit age?", 0);
                          FocusScope.of(context).requestFocus(_nodes.elementAt(0));
                        });
                      },
                      // Sets the field back to read-only
                      onSubmitted: (d){setState(() {
                        _ReadOnlybools[0] = true;
                        postUser(jsonEncode(<String, int>{
                          'age': int.parse(_ageController.text),
                          'height': SessionHeight,
                          'weight': SessionWeight,
                        }));
                        SessionAge = int.parse(_ageController.text);
                      });
                      },
                      focusNode: _nodes.elementAt(0),
                      readOnly: _ReadOnlybools.elementAt(0),
                      controller: _ageController,
                      decoration: InputDecoration(
                        suffixIcon: Icon(Icons.create),
                        border: OutlineInputBorder(),
                        labelText: 'Age',
                      )
                  )
              ),
              Padding(
                  padding: EdgeInsets.fromLTRB(0, 0, 0, 10),
                  child: TextField(
                      onTap: () {
                         setState(() {
                          _showDialog("Edit height?", 1);
                          FocusScope.of(context).requestFocus(_nodes.elementAt(1));
                        });
                      },
                      // Sets the field back to read-only
                      onSubmitted: (e){setState(() {
                        _ReadOnlybools[1] = true;
                        postUser(jsonEncode(<String, int>{
                          'age': SessionAge,
                          'height': int.parse(_heightController.text),
                          'weight': SessionWeight,
                        }));
                        SessionHeight = int.parse(_heightController.text);
                      });
                      },
                      focusNode: _nodes.elementAt(1),
                      readOnly: _ReadOnlybools.elementAt(1),
                      controller: _heightController,
                      decoration: InputDecoration(
                        suffixIcon: Icon(Icons.create),
                        border: OutlineInputBorder(),
                        labelText: 'Height',
                      )
                  )
              ),
                Padding(
                  padding: EdgeInsets.fromLTRB(0, 0, 0, 10),
                  child: TextField(
                      onTap: () {
                        setState(() {
                          _showDialog("Edit weight?", 2);
                          FocusScope.of(context).requestFocus(_nodes.elementAt(2));
                        });
                      },
                      // Sets the field back to read-only
                      onSubmitted: (e){setState(() {
                        _ReadOnlybools[2] = true;
                        postUser(jsonEncode(<String, int>{
                          'age': SessionAge,
                          'height': SessionHeight,
                          'weight': int.parse(_weightController.text),
                        }));
                        SessionWeight = int.parse(_weightController.text);
                      });
                      },
                      focusNode: _nodes.elementAt(2),
                      readOnly: _ReadOnlybools.elementAt(2),
                      controller: _weightController,
                      decoration: InputDecoration(
                        suffixIcon: Icon(Icons.create),
                        border: OutlineInputBorder(),
                        labelText: 'Weight',
                      )
                  )
                ),
            ],
        )
      ])
    );
  }
}