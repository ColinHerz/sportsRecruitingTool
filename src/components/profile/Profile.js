import apiCall from "../../api/api.js";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import React, { useState, useEffect } from "react";

import "./profile.scss";

const Profile = props => {
	const [addingBag, setAddingBag] = useState(false);
	const [newBagName, setNewBagName] = useState(``);

	const [bags, setBags] = useState([]);
	const [updateBags, setUpdateBags] = useState(false);

	const [updateUser, setUpdateUser] = useState(true);
	const [userData, setUserData] = useState({});
	const [newUserData, setNewUserData] = useState({});
	const [updatingUserData, setUpdatingUserData] = useState(false);

	const [verificationSent, setVerificationSent] = useState(false);

	const [showError, setShowError] = useState(false);

	useEffect(() => {
		if (updateBags) {
			apiCall(
				{
					endpoint: `/golf/getAllGolfBags`,
					type: `GET`
				},
				data => {
					if (data.status !== 200) {
						setShowError(true);
						setBags([]);
						return;
					}

					setBags(JSON.parse(data.response));
				},
				() => {
					setShowError(true);
					setBags([]);
				}
			);

			setUpdateBags(false);
		}
	}, [updateBags]);

	useEffect(() => {
		if (updateUser) {
			userInfo();
			setUpdateUser(false);
		}
	}, [updateUser]);

	// API call methods

	const addBag = event => {
		event.preventDefault();

		apiCall(
			{
				endpoint: `/golf/createGolfBag`,
				type: `POST`,
				body: {
					bagName: newBagName
				}
			},
			data => {
				if (data.status !== 200) {
					setShowError(true);
				}

				setNewBagName(``);
				setAddingBag(false);
				setUpdateBags(true);
			},
			() => {
				setShowError(true);
			}
		);
	};

	const deleteBag = event => {
		event.preventDefault();

		const id = event.target.parentElement.id;

		apiCall(
			{
				endpoint: `/golf/deleteGolfBag`,
				type: `POST`,
				body: {
					golfBag: id
				}
			},
			data => {
				if (data.status !== 200) {
					setShowError(true);
				}

				setUpdateBags(true);
			},
			() => {
				setShowError(true);
			}
		);
	};

	const userInfo = () => {
		apiCall(
			{
				endpoint: `/users/detail/get`,
				type: `GET`,
			},
			data => {
				if (data.status !== 200) {
					setShowError(true);
					return;
				}

				const response = data.response;
				let userData;

				if (response === ``) {
					userData = {
						age: ``,
						height: ``,
						weight: ``
					};
				}
				else {
					userData = JSON.parse(data.response);

					if (userData.age === undefined) {
						userData.age = ``;
					}

					if (userData.height === undefined) {
						userData.age = ``;
					}

					if (userData.weight === undefined) {
						userData.age = ``;
					}
				}

				setUserData(userData);
				setNewUserData(userData);
			},
			() => {
				setShowError(true);
			}
		);
	};

	const sendUserData = () => {
		apiCall(
			{
				endpoint: `/users/detail/update`,
				type: `POST`,
				body: newUserData
			},
			data => {
				if (data.status !== 200) {
					setShowError(true);
					return;
				}

				setUpdateUser(true);
			},
			() => {
				setShowError(true);
			}
		);
	};

	const resendVerification = event => {
		event.preventDefault();

		apiCall(
			{
				endpoint: `/emails/resendVerificationEmail`,
				type: `POST`,
				body: {
					recieverEmail: props.user.email
				}
			},
			data => {
				if (data.status !== 200) {
					setShowError(true);
					return;
				}

				setVerificationSent(true);
			},
			() => {
				setShowError(true);
			}
		);
	};

	// Info handling and transformation functions

	const showEditProfile = event => {
		event.preventDefault();

		setUpdatingUserData(true);
	};

	const updateAge = event => {
		event.preventDefault();

		const data = Object.assign({},newUserData);
		data.age = parseInt(event.target.value);

		if (isNaN(data.age)) {
			return;
		}

		setNewUserData(data);
	};

	const updateWeight = event => {
		event.preventDefault();

		const data = Object.assign({},newUserData);
		data.weight = parseInt(event.target.value);

		if (isNaN(data.weight)) {
			return;
		}

		setNewUserData(data);
	};

	const updateHeight = event => {
		event.preventDefault();

		const data = Object.assign({},newUserData);
		data.height = parseInt(event.target.value);

		if (isNaN(data.height)) {
			return;
		}

		setNewUserData(data);
	};

	const cancelUpdateInfo = event => {
		event.preventDefault();

		setUpdatingUserData(false);
	};

	const showAddingBag = event => {
		event.preventDefault();

		setAddingBag(true);
	};

	const cancelAddBag = event => {
		event.preventDefault();

		setNewBagName(``);
		setAddingBag(false);
	};

	const handleBagNameChange = event => {
		event.preventDefault();

		setNewBagName(event.target.value);
	};

	return (
		<main id="profile">
			{
				showError ?
					<p id="error-message">Something went wrong. Please try again latter.</p>:
					null
			}

			<section id="login-info">
				<h2>{`${props.user.firstname} ${props.user.lastname}`}</h2>

				<p>{`${props.user.email}`}{
					props.user.isVerified ?
						<span id="verified"> Verified</span>:
						<span id="not-verified"> Not Verified</span>
				}</p>

				<button onClick={resendVerification}>Click to resend verification email.</button>

				{
					verificationSent ?
						<p id="email-confirmation">Email sent.</p>:
						null
				}
			</section>

			<section id="personal-info">
				<ul>
					<li><span>Age: </span>{userData.age}</li>
					<li><span>Height: </span>{userData.height}{userData.height !== `` ? `"`:null}</li>
					<li><span>Weight: </span>{userData.weight}{userData.weight !== `` ? ` lbs`:null}</li>
				</ul>

				<button id="edit-info" onClick={showEditProfile}>Edit Profile</button>

				{
					updatingUserData ?
						<div id="updated-info">
							<label className="userDetails">
								Height (inches):
								<input
									type="text"
									value={newUserData.height}
									onChange={updateHeight}
								/>
							</label>

							<label className="userDetails">
								Weight (pounds):
								<input
									type="text"
									value={newUserData.weight}
									onChange={updateWeight}
								/>
							</label>

							<label className="userDetails">
								Age (years):
								<input
									type="text"
									value={newUserData.age}
									onChange={updateAge}
								/>
							</label>

							<div>
								<button onClick={sendUserData}>Change</button>
								<button onClick={cancelUpdateInfo}>Cancel</button>
							</div>
						</div>:
						null
				}
			</section>

			<section id="bags">
				<h3>Bags</h3>

				<ul>
					{
						bags.map(bag => {
							return (
								<li key={bag._id} id={bag._id}>
									<Link
										to={`/profile/bag/${bag._id}/`}
									>
										{bag.bagName}
									</Link>

									<button onClick={deleteBag}>Delete</button>
								</li>
							);
						})
					}
				</ul>

				{
					addingBag ?
						<div id="adding-bag-dialogue">
							<label>
								Bag Name:
								<input
									type="text"
									id="bag-name"
									value={newBagName}
									onChange={handleBagNameChange}
								/>
							</label>

							<button onClick={addBag}>Add</button>
							<button onClick={cancelAddBag}>Cancel</button>
						</div>:
						null
				}

				<button
					id="add-bag"
					onClick={showAddingBag}
				>
					Add Bag
				</button>
			</section>
		</main>
	);
};

Profile.propTypes = {
	user: PropTypes.object.isRequired
};

export default Profile;
