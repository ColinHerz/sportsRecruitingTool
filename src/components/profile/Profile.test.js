/* eslint-env jest */
import { MemoryRouter } from "react-router-dom";
import React from "react";
import renderer from "react-test-renderer";

import Profile from "./Profile.js";

const genProps = isVerified => {
	return {
		user: {
			id: `garbage`,
			email: `john.smith@yahoo.net`,
			firstname: `John`,
			lastname: `Smith`,
			isVerified: isVerified
		}
	};
};

beforeEach(() => {
	jest.resetModules();
});

describe(`Rendering`, () => {
	it(`Renders correctly when not verified`, () => {
		const props = genProps(false);

		const component = renderer.create(
			<MemoryRouter>
				<Profile {...props} />
			</MemoryRouter>
		);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it(`Renders correctly when verified`, () => {
		const props = genProps(true);

		const component = renderer.create(
			<MemoryRouter>
				<Profile {...props} />
			</MemoryRouter>
		);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
