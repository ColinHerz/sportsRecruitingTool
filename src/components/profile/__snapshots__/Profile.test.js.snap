// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Renders correctly when not verified 1`] = `
<main
  id="profile"
>
  <section
    id="login-info"
  >
    <h2>
      John Smith
    </h2>
    <p>
      john.smith@yahoo.net
      <span
        id="not-verified"
      >
         Not Verified
      </span>
    </p>
    <button
      onClick={[Function]}
    >
      Click to resend verification email.
    </button>
  </section>
  <section
    id="personal-info"
  >
    <ul>
      <li>
        <span>
          Age: 
        </span>
      </li>
      <li>
        <span>
          Height: 
        </span>
        "
      </li>
      <li>
        <span>
          Weight: 
        </span>
         lbs
      </li>
    </ul>
    <button
      id="edit-info"
      onClick={[Function]}
    >
      Edit Profile
    </button>
  </section>
  <section
    id="bags"
  >
    <h3>
      Bags
    </h3>
    <ul />
    <button
      id="add-bag"
      onClick={[Function]}
    >
      Add Bag
    </button>
  </section>
  <section
    id="matches"
  >
    <h3>
      Matches
    </h3>
    <ul />
    <a
      href="/match/create/"
      id="create-match-link"
      onClick={[Function]}
    >
      Create Match
    </a>
  </section>
</main>
`;

exports[`Rendering Renders correctly when verified 1`] = `
<main
  id="profile"
>
  <section
    id="login-info"
  >
    <h2>
      John Smith
    </h2>
    <p>
      john.smith@yahoo.net
      <span
        id="verified"
      >
         Verified
      </span>
    </p>
    <button
      onClick={[Function]}
    >
      Click to resend verification email.
    </button>
  </section>
  <section
    id="personal-info"
  >
    <ul>
      <li>
        <span>
          Age: 
        </span>
      </li>
      <li>
        <span>
          Height: 
        </span>
        "
      </li>
      <li>
        <span>
          Weight: 
        </span>
         lbs
      </li>
    </ul>
    <button
      id="edit-info"
      onClick={[Function]}
    >
      Edit Profile
    </button>
  </section>
  <section
    id="bags"
  >
    <h3>
      Bags
    </h3>
    <ul />
    <button
      id="add-bag"
      onClick={[Function]}
    >
      Add Bag
    </button>
  </section>
  <section
    id="matches"
  >
    <h3>
      Matches
    </h3>
    <ul />
    <a
      href="/match/create/"
      id="create-match-link"
      onClick={[Function]}
    >
      Create Match
    </a>
  </section>
</main>
`;
